lib/vtls/darwinssl.c

```diff
@@ -1996,6 +2025,127 @@ static int verify_cert(const char *cafile, struct Curl_easy *data,
   }
 }
 
+#ifdef DARWIN_SSL_PINNEDPUBKEY
+static CURLcode pkp_pin_peer_pubkey(struct SessionHandle *data,
+                                    SSLContextRef ctx,
+                                    const char *pinnedpubkey)
+{  /* Scratch */
+  size_t pubkeylen, realpubkeylen, derHeaderLength = 24;
+  unsigned char *pubkey = NULL, *realpubkey = NULL, *derHeader = NULL;
+  CFDataRef publicKeyBits = NULL;
+
+  /* Result is returned to caller */
+  CURLcode result = CURLE_SSL_PINNEDPUBKEYNOTMATCH;
+
+  /* if a path wasn't specified, don't pin */
+  if(!pinnedpubkey)
+    return CURLE_OK;
+
+
+  if(!ctx)
+    return result;
+
+  do {
+    SecTrustRef trust;
+    OSStatus ret = SSLCopyPeerTrust(ctx, &trust);
+    if(ret != noErr || trust == NULL)
+        break;
+
+    SecKeyRef keyRef;
+
+#ifdef DARWIN_SSL_PINNEDPUBKEY_V1
+
+    /* 0 is leaf certificate, this is root:
+     * CFIndex certificateChainLen = SecTrustGetCertificateCount(trust);
+     */
+    SecCertificateRef cert = SecTrustGetCertificateAtIndex(trust, 0);
+    SecPolicyRef policy = SecPolicyCreateBasicX509();
+
+    SecTrustRef tempTrust;
+    SecTrustCreateWithCertificates(cert, policy, &tempTrust);
+    SecTrustEvaluate(tempTrust, NULL);
+    keyRef = SecTrustCopyPublicKey(tempTrust);
```

Comment:
Wow, this is a lot of work to do the equivalent of calling SecTrustCopyPublicKey on the original trust object. Why, exactly, are we doing this?